bin/wasm-node/rust/src/platform.rs

```diff
@@ -106,183 +104,463 @@ impl smoldot_light_base::Platform for Platform {
-                this.messages_queue.pop_front();
-                this.messages_queue_first_offset = 0;
+            // Move the next buffer from `STATE` into `read_buffer`.
+            let mut lock = STATE.try_lock().unwrap();
+            let stream = lock.streams.get_mut(stream_id).unwrap();
+            if let Some(msg) = stream.messages_queue.pop_front() {
+                read_buffer.buffer = msg;
+                read_buffer.buffer_first_offset = 0;
+            } else {
+                panic!() // User has passed more bytes than the size of the read buffer.
             }
-        } else {
-            assert_eq!(bytes, 0);
-        };
+        }
     }
 
-    fn send(connection: &mut Self::Stream, data: &[u8]) {
-        unsafe {
-            let this = Pin::get_unchecked_mut(connection.as_mut());
+    fn send(StreamWrapper((connection_id, stream_id), _): &mut Self::Stream, data: &[u8]) {
+        let mut lock = STATE.try_lock().unwrap();
+        let stream = lock.streams.get_mut(&(*connection_id, *stream_id)).unwrap();
 
-            // Connection might have been closed, but API user hasn't detected it yet.
-            if this.closed_message.is_some() {
-                return;
-            }
+        if stream.closed {
+            return;
+        }
 
-            TOTAL_BYTES_SENT.fetch_add(data.len(), Ordering::Relaxed);
+        TOTAL_BYTES_SENT.fetch_add(data.len(), Ordering::Relaxed);
 
+        unsafe {
             bindings::stream_send(
-                this.id.unwrap(),
-                0,
+                *connection_id,
+                *stream_id,
                 u32::try_from(data.as_ptr() as usize).unwrap(),
                 u32::try_from(data.len()).unwrap(),
             );
         }
     }
 }
 
-/// Connection connected to a target.
-pub(crate) struct Connection {
-    /// If `Some`, [`bindings::connection_close`] must be called. Set to a value after
-    /// [`bindings::connection_new`] returns success.
-    id: Option<u32>,
-    /// True if [`bindings::connection_open_single_stream`] has been called.
-    open: bool,
-    /// `Some` if [`bindings::connection_closed`] has been called.
-    closed_message: Option<String>,
-    /// List of messages received through [`bindings::stream_message`]. Must never contain
-    /// empty messages.
-    messages_queue: VecDeque<Box<[u8]>>,
-    /// Position of the read cursor within the first element of [`Connection::messages_queue`].
-    messages_queue_first_offset: usize,
-    /// Event notified whenever one of the fields above is modified.
-    something_happened: event_listener::Event,
-    /// Prevents the [`Connection`] from being unpinned.
-    _pinned: marker::PhantomPinned,
-}
+pub(crate) struct StreamWrapper((u32, u32), ReadBuffer);
+
+impl Drop for StreamWrapper {
+    fn drop(&mut self) {
+        let mut lock = STATE.try_lock().unwrap();
+
+        let connection = lock.connections.get_mut(&self.0 .0).unwrap();
```

Comment:
format the code?

```suggestion
        let connection = lock.connections.get_mut(&self.0.0).unwrap();
```

Reply:
Rustfmt formats this code as `self.0 .0`.
In the past the compiler treated this as `self.(0.0)`, which was a compilation error. This has been fixed, but rustfmt has apparently not been updated.